### PR TITLE
Batchnorm bwd cpu ref

### PR DESCRIPTION
**Pull request by SamuelReeder**
## Proposed changes

Batchnorm bwd cpu ref implementation and tests.

## Checklist

Please put an `x` into the boxes that apply. You can also fill these out after creating the PR. If you're not sure, please don't hesitate to ask.

- [x] I have added automated tests relevant to the introduced functionality
- [x] I have sufficient test coverage for the changes, and code coverage hasn't decreased as a result of my PR
- [x] I have ran the tests, and they are all passing locally
- [x] I have ran the tests with ASAN, and they are all passing locally
- [ ] I have added relevant documentation for the changes
- [ ] I have removed the stale documentation which is no longer relevant after this pull request
- [x] I have ran `make format` & `make check_format` to ensure the changes have been formatted

### CODE REVIEW

File: sdk/tests/test_utilities/test_cpu_fp_reference_implementation.cpp [hunk 1]
```diff
@@ -13,87 +13,281 @@ using namespace hipdnn_sdk::reference_test_utilities;
...
 
     Cpu_fp_reference_implementation<float, float, float> ref_impl;
 
     ref_impl.batchnorm_fwd_inference(
         input_tensor, scale_tensor, bias_tensor, mean_tensor, variance_tensor, output_tensor, 1e-5);
 }
+
+TEST(CpuFpReferenceImplementation, BatchnormBwdFloatUsage)
+{
+    SKIP_IF_NO_DEVICES();
```
Comment (SamuelReeder) [inline sdk/tests/test_utilities/test_cpu_fp_reference_implementation.cpp:103]: I have the `SKIP_IF_NO_DEVICES()` here because the rest of the ref tests have them, but I don't get why we use them here since it doesn't involve a device.

File: plugins/miopen_legacy_plugin/integration_tests/batchnorm_fwd_inference_integration_test.cpp [hunk 2]
```diff
@@ -284,17 +284,16 @@ class Batchnorm_forward_inference_integration_test
...
+        {.n = 1, .c = 3, .h = 14, .w = 14} // {.n = 2, .c = 3, .h = 14, .w = 14},
+        // {.n = 64, .c = 3, .h = 14, .w = 14},
+        // {.n = 64, .c = 256, .h = 14, .w = 14},
+        // {.n = 64, .c = 256, .h = 28, .w = 28},
+        // {.n = 64, .c = 256, .h = 56, .w = 56},
+        // {.n = 64, .c = 512, .h = 14, .w = 14},
+        // {.n = 64, .c = 512, .h = 28, .w = 28},
+        // {.n = 64, .c = 512, .h = 7, .w = 7},
+        // {.n = 64, .c = 64, .h = 112, .w = 112},
+        // {.n = 64, .c = 64, .h = 56, .w = 56},
```
Comment (SamuelReeder) [inline plugins/miopen_legacy_plugin/integration_tests/batchnorm_fwd_inference_integration_test.cpp]: Oops will uncomment.

File: sdk/tests/test_utilities/test_cpu_fp_reference_implementation.cpp [hunk 3]
```diff
@@ -13,87 +13,281 @@ using namespace hipdnn_sdk::reference_test_utilities;
...
+                           x_tensor,
+                           mean_tensor,
+                           inv_variance_tensor,
+                           scale_tensor,
+                           dx_tensor,
+                           dscale_tensor,
+                           dbias_tensor);
+}
+
+TEST(CpuFpReferenceImplementation, BatchnormBwdSanityValidation)
```
Comment (SamuelReeder) [inline sdk/tests/test_utilities/test_cpu_fp_reference_implementation.cpp:226]: This is no golden data but I wanted a clear sanity check. Can remove if its pointless to keep.

File: sdk/tests/test_utilities/test_cpu_fp_reference_implementation.cpp (hunk 1 above)
Comment (mousdahl-amd) [reply sdk/tests/test_utilities/test_cpu_fp_reference_implementation.cpp:103]: Have you tried running ASAN without it?  If it works, that's great.

File: sdk/tests/test_utilities/test_cpu_fp_reference_implementation.cpp (hunk 3 above)
Comment (mousdahl-amd) [reply sdk/tests/test_utilities/test_cpu_fp_reference_implementation.cpp:226]: I like it!

Comment (adickin-amd) [reply sdk/tests/test_utilities/test_cpu_fp_reference_implementation.cpp:226]: I think this is great! Very clear comments to show how the expected results are calculated.

**Review by mousdahl-amd — APPROVED** [review]
LGTM, remove SKIP_IF_NO_DEVICES if possible

File: sdk/include/hipdnn_sdk/test_utilities/cpu_fp_reference_implementation.hpp [hunk 4]
```diff
@@ -83,6 +83,103 @@ class Cpu_fp_reference_implementation : public Reference_implementation_interfac
...
+
+            // Calculate dot product for (x - mean) * channel_inv_variance * dy and ∑ dy for this channel
+            Mean_variance_data_type dot_product = 0;
+            Mean_variance_data_type sum_dy = 0;
+
+            for(int row = 0; row < height; row++)
+            {
+                for(int column = 0; column < width; column++)
+                {
+                    for(int bidx = 0; bidx < n_batches; bidx++)
```
Comment (SamuelReeder) [inline sdk/include/hipdnn_sdk/test_utilities/cpu_fp_reference_implementation.hpp]: Would it be better to go batch -> row -> col instead for mem locality?

File: sdk/tests/test_utilities/test_cpu_fp_reference_implementation.cpp (hunk 3 above)
Comment (SamuelReeder) [reply sdk/tests/test_utilities/test_cpu_fp_reference_implementation.cpp:226]: Ooh, got same results in pytorch `batch_norm` too:
```
dbias:  1.000000000000000
dscale: 0.447211806656309

dx tensor:
tensor([[[[-2.1466e-06, -7.1553e-07],
          [ 7.1553e-07,  2.1466e-06]]]], dtype=torch.float64)
```

File: sdk/tests/test_utilities/test_cpu_fp_reference_implementation.cpp (hunk 1 above)
Comment (SamuelReeder) [reply sdk/tests/test_utilities/test_cpu_fp_reference_implementation.cpp:103]: It still fails so I'll keep them, must be `hipHostMalloc` or something?

Comment (mousdahl-amd) [reply sdk/tests/test_utilities/test_cpu_fp_reference_implementation.cpp:103]: That makes sense, and reminds me.  Adam was suggesting we change Migratable_memory to use plain old malloc for host allocations, and that would allow us to get rid of the SKIP_IF_NO_DEVICES() checks for these CPU centric tests.  I'm doing a rework of Migratable_memory in my issue, so I'll add a note to try that change and then come back and remove these.

Comment (mousdahl-amd) [reply sdk/tests/test_utilities/test_cpu_fp_reference_implementation.cpp:103]: Okay, note added to my story.  It's now my problem!

File: sdk/include/hipdnn_sdk/test_utilities/cpu_fp_reference_implementation.hpp (hunk 4 above)
Comment (BrianHarrisonAMD) [reply sdk/include/hipdnn_sdk/test_utilities/cpu_fp_reference_implementation.hpp]: Yea

Comment (BrianHarrisonAMD) [reply sdk/include/hipdnn_sdk/test_utilities/cpu_fp_reference_implementation.hpp]: Best is to go by smallest stride between the 3.

Which will be batch -> row -> col for NCHW & NHWC.

File: sdk/tests/test_utilities/test_cpu_fp_reference_implementation.cpp (hunk 1 above)
Comment (BrianHarrisonAMD) [reply sdk/tests/test_utilities/test_cpu_fp_reference_implementation.cpp:103]: Yea any hip call is problematic even hipHostMalloc.

File: sdk/include/hipdnn_sdk/test_utilities/cpu_fp_reference_implementation.hpp [hunk 5]
```diff
@@ -83,6 +83,103 @@ class Cpu_fp_reference_implementation : public Reference_implementation_interfac
...
+                       const Tensor& invVariance,
+                       const Tensor& scale,
+                       Tensor& dx,
+                       Tensor& dscale,
+                       Tensor& dbias) override
+    {
+        if(x.dims().size() != 4)
+        {
+            throw std::runtime_error("Batchnorm backward requires a 4D tensor.");
+        }
```
Comment (BrianHarrisonAMD) [inline sdk/include/hipdnn_sdk/test_utilities/cpu_fp_reference_implementation.hpp:99]: I guess since it's the SDK we can't use our exception class?

File: sdk/include/hipdnn_sdk/test_utilities/cpu_fp_reference_implementation.hpp [hunk 6]
```diff
@@ -83,6 +83,103 @@ class Cpu_fp_reference_implementation : public Reference_implementation_interfac
         output.memory().mark_host_modified(); // Mark output memory as modified on host
     }
 
+    void batchnorm_bwd(const Tensor& dy,
+                       const Tensor& x,
+                       const Tensor& mean,
+                       const Tensor& invVariance,
```
Comment (BrianHarrisonAMD) [inline sdk/include/hipdnn_sdk/test_utilities/cpu_fp_reference_implementation.hpp]: Does this not flag a format issue?

I thought it needed to be `inv_variance`

**Review by BrianHarrisonAMD — APPROVED** [review]
Looks generally good to me.

Can improve the loop ordering, and maybe double check formatting since I am surprised it's not flagging.

File: sdk/include/hipdnn_sdk/test_utilities/cpu_fp_reference_implementation.hpp (hunk 5 above)
Comment (SamuelReeder) [reply sdk/include/hipdnn_sdk/test_utilities/cpu_fp_reference_implementation.hpp:99]: Yeah, there's currently no sdk exception class. But there could be!

File: sdk/include/hipdnn_sdk/test_utilities/cpu_fp_reference_implementation.hpp (hunk 6 above)
Comment (SamuelReeder) [reply sdk/include/hipdnn_sdk/test_utilities/cpu_fp_reference_implementation.hpp]: You're right, but it doesn't flag either way. It is formatting these files, but I don't think all clang tidy rules are being enforced on them... I'll see if its a quick fix.

Comment (SamuelReeder) [reply sdk/include/hipdnn_sdk/test_utilities/cpu_fp_reference_implementation.hpp]: Was a missing `clang_tidy_check(...)`. Turning it on reveals new issues, so I'll either fix here or in another PR depending on how big.
